Materialized view

Gor027 opened the pull request:
## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [ ] I have enabled appropriate tests in `.github/workflows/build.yaml` in `gtest_filter`.

This PR is dependent upon #72 

Adds support to fetch materialized views from cluster metadata and implementation for functions to return fields of a certain view in the metadata. Additionally, it adds support to iterate over materialized views.

Lorak-mmk commented:
Please rebase on current version of #72

Lorak-mmk commented:
Please rebase on current master - it's a bit difficult to review with those unrelated changes from #72

Lorak-mmk commented on scylla-rust-wrapper/src/session.rs:
```diff
@@ -774,3 +774,204 @@ pub unsafe extern "C" fn cass_column_meta_type(
+
+    match &table_meta.views {
+        Some(table_views) => match table_views.get(view_name) {
+            Some(view_meta) => Arc::as_ptr(view_meta) as *const CassMaterializedViewMeta,
```
Again, why is this cast needed?

Gor027 replied:
It is not necessary and can be safely removed.

Lorak-mmk reviewed (changes requested):
I don't like that all the metadata methods are in session.rs - they should have their own file, as they are not methods on the CassSesson object.

Gor027 commented on scylla-rust-wrapper/src/user_type.rs:
```diff
@@ -73,13 +73,13 @@ impl From<&CassUserType> for CqlValue {
-    let data_type = clone_arced(data_type_raw);
+    let data_type = ptr_to_ref(data_type_raw);
 
-    match &*data_type {
+    match data_type {
         CassDataType::UDT(udt_data_type) => {
             let field_values = vec![None; udt_data_type.field_types.len()];
             Box::into_raw(Box::new(CassUserType {
-                data_type,
+                data_type: Arc::new(CassDataType::UDT(udt_data_type.clone())),
```
Yes, this is an unnecessary change, the commit should be dropped.

Gor027 commented on scylla-rust-wrapper/src/user_type.rs:
```diff
@@ -73,13 +73,13 @@ impl From<&CassUserType> for CqlValue {
 pub unsafe extern "C" fn cass_user_type_new_from_data_type(
     data_type_raw: *const CassDataType,
 ) -> *mut CassUserType {
-    let data_type = clone_arced(data_type_raw);
+    let data_type = ptr_to_ref(data_type_raw);
 
-    match &*data_type {
+    match data_type {
         CassDataType::UDT(udt_data_type) => {
             let field_values = vec![None; udt_data_type.field_types.len()];
             Box::into_raw(Box::new(CassUserType {
-                data_type,
+                data_type: Arc::new(CassDataType::UDT(udt_data_type.clone())),
                 field_values,
             }))
         }
```
These changes survived after a rebase as there were memory corruption issues with this part of the code. As we have debugged and solved the problem, the changes will be reverted.

Gor027 commented on scylla-rust-wrapper/src/session.rs:
```diff
@@ -774,3 +774,204 @@ pub unsafe extern "C" fn cass_column_meta_type(
+    let view_name = ptr_to_cstr_n(view, view_length).unwrap();
+
+    match keyspace_meta.views.get(view_name) {
+        Some(view_meta) => Arc::as_ptr(view_meta) as *const CassMaterializedViewMeta,
```
The type returned by `Arc::as_ptr` should be `*const CassMaterializedVIewMeta`, I added the cast here to explicitly mark in the code what is returned from the function. I guess anyway the compiler does some optimization and removes the cast here.

Lorak-mmk replied:
I think the cast makes it less readable, can you remove it?

Gor027 commented on scylla-rust-wrapper/src/session.rs:
```diff
@@ -49,6 +51,13 @@ pub struct CassTableMeta {
+pub struct CassMaterializedViewMeta {
+    pub name: String,
+    pub view_metadata: CassTableMeta,
+    pub base_table: Arc<RefCell<CassTableMeta>>,
```
Reference to the base table of a materialized view should be Weak, you are right. I will add some tests to ensure we avoid memory leaks.

Lorak-mmk commented on scylla-rust-wrapper/src/session.rs:
```diff
@@ -39,7 +39,8 @@ pub struct CassKeyspaceMeta {
     // User defined type name to type
     pub user_defined_type_data_type: HashMap<String, Arc<CassDataType>>,
-    pub tables: HashMap<String, CassTableMeta>,
+    pub tables: HashMap<String, Arc<CassTableMeta>>,
+    pub views: HashMap<String, Arc<CassMaterializedViewMeta>>,
```
This is not that important, but why metadata-related structures use `Arc` and not just `Rc`? cppdriver only requires `CassSession` and `CassFuture` to be thread safe. Do we do something similar (stronger thread-safety than cppdriver requires) in other places?
Here, using `Arc` instead of `Rc` allows to call `cass_schema_meta_free` on a different thread than `cass_session_get_schema_meta` was called from (those are the only places where the counters are modified) - that doesn't seem like a big benefit.

Lorak-mmk reviewed (changes requested):
I don't quite like how are the structs / functions places between `session.rs` and `metadata.rs`. Before this PR we had no `metadata.rs` file, so everything was put into `session.rs.` Now that we have it, we can fix this situation.

I'd propose the following:
- Leave `query_results.rs` as it is - meaning that functions accepting or returning `CassIterator` go there.
- Move metadata structs to `metadata.rs` - `CassKeyspaceMeta_`, `CassKeyspaceMeta`, `CassTableMeta_`, `CassTableMeta`, `CassMaterializedViewMeta_`, `CassMaterializedViewMeta`, `CassColumnMeta`, `CassSchemaMeta_`, `CassSchemaMeta`.
- Move metadata functions that don't touch session into `metadata.rs` - everything from `cass_schema_meta_free` downward.
- `create_table_metadata` should probably also go into `metadata.rs`, but I'm not 100% sure about it.

Gor027 commented on scylla-rust-wrapper/src/session.rs:
```diff
@@ -464,40 +475,34 @@ pub unsafe extern "C" fn cass_session_get_schema_meta(
+                    view_metadata: cass_view_table_meta,
+                    base_table: Weak::new(),
+                };
+                cass_view_meta.base_table = Arc::downgrade(&cass_table_meta);
```
The initialization of `cass_table_meta` with `Arc::new_cyclic` allowed to avoid `Arc::downgrade` and to make `cass_view_meta` immutable.

Gor027 commented on scylla-rust-wrapper/src/session.rs:
```diff
@@ -49,6 +50,15 @@ pub struct CassTableMeta {
     pub columns_metadata: HashMap<String, CassColumnMeta>,
     pub partition_keys: Vec<String>,
     pub clustering_keys: Vec<String>,
+    pub views: std::sync::RwLock<Option<HashMap<String, Arc<CassMaterializedViewMeta>>>>,
```
The purpose of `Option` was to differentiate uninitialized `views` from empty `views` for better error handling, however, it seemed to be useless, so I removed `Option`.

Gor027 commented on scylla-rust-wrapper/src/session.rs:
```diff
@@ -49,6 +50,15 @@ pub struct CassTableMeta {
     pub columns_metadata: HashMap<String, CassColumnMeta>,
     pub partition_keys: Vec<String>,
     pub clustering_keys: Vec<String>,
+    pub views: std::sync::RwLock<Option<HashMap<String, Arc<CassMaterializedViewMeta>>>>,
```
Initialization with `Arc::new_cyclic` allowed to get rid of `RwLock` and keep the `views` type as simple as `HashMap<String, Arc<CassMaterializedViewMeta>>`.

Gor027 commented:
> I don't quite like how are the structs / functions places between `session.rs` and `metadata.rs`. Before this PR we had no `metadata.rs` file, so everything was put into `session.rs.` Now that we have it, we can fix this situation.
> 
> I'd propose the following:
> 
> * Leave `query_results.rs` as it is - meaning that functions accepting or returning `CassIterator` go there.
> * Move metadata structs to `metadata.rs` - `CassKeyspaceMeta_`, `CassKeyspaceMeta`, `CassTableMeta_`, `CassTableMeta`, `CassMaterializedViewMeta_`, `CassMaterializedViewMeta`, `CassColumnMeta`, `CassSchemaMeta_`, `CassSchemaMeta`.
> * Move metadata functions that don't touch session into `metadata.rs` - everything from `cass_schema_meta_free` downward.
> * `create_table_metadata` should probably also go into `metadata.rs`, but I'm not 100% sure about it.
> 
> What do you think?

As you suggested above, I moved all metadata-related structures and functions to `metadata.rs`. Importing the `CassColumnType` enum is also moved to `metadata.rs`, so I think it is better to have `create_table_metadata` also in that file rather than in `session.rs`.

Lorak-mmk commented on scylla-rust-wrapper/src/metadata.rs:
```diff
@@ -0,0 +1,502 @@
+    pub column_kind: CassColumnType,
+}
+
+pub type CassSchemaMeta_ = &'static CassSchemaMeta;
+
+pub struct CassSchemaMeta {
+    pub keyspaces: HashMap<String, CassKeyspaceMeta>,
+}
```
Nit / subjective: Move them above `CassKeyspaceMeta_`

Gor027 replied:
Done.

Lorak-mmk commented on scylla-rust-wrapper/src/metadata.rs:
```diff
@@ -0,0 +1,502 @@
+        .user_defined_type_data_type
+        .get(user_type_name)
+    {
+        Some(udt) => Arc::into_raw(udt.clone()) as *const CassDataType,
```
Why is there a clone? I see that cpp-driver doesn't clone this, so we'd be incompatible (in this version user would have to explicitly free this udt, in contrary to cpp-driver).

Gor027 replied:
The `udt` object is of type `&Arc<CassDataType>` and `Arc::into_raw` expects argument by value rather than by reference, so I think a clone is necessary here.

Gor027 replied:
Ok, there is the `Arc::as_ptr` function which helps to avoid the clone. Will fix it.

Lorak-mmk commented on scylla-rust-wrapper/src/metadata.rs:
```diff
@@ -0,0 +1,502 @@
+    let keyspace_meta = metadata.keyspaces.get(keyspace);
+
+    match keyspace_meta {
+        Some(meta) => meta,
```
`meta` is a reference, so maybe add an explicit cast here?

Gor027 replied:
Done.

Lorak-mmk commented on scylla-rust-wrapper/src/metadata.rs:
```diff
@@ -0,0 +1,502 @@
+    let view_name = ptr_to_cstr_n(view, view_length).unwrap();
+
+    match table_meta.views.get(view_name) {
+        Some(view_meta) => Arc::as_ptr(view_meta) as *const CassMaterializedViewMeta,
```
Remove the cast (as_ptr already returns correct type)

Gor027 replied:
Done.

Lorak-mmk commented on scylla-rust-wrapper/src/metadata.rs:
```diff
@@ -0,0 +1,502 @@
+    let table_meta = ptr_to_ref(table_meta);
+
+    match table_meta.views.iter().nth(index as usize) {
+        Some(view_meta) => Arc::as_ptr(view_meta.1) as *const CassMaterializedViewMeta,
```
Remove the cast (as_ptr already returns correct type)

Gor027 replied:
Done.

Lorak-mmk replied:
No it's not

Gor027 replied:
The cast perhaps appeared after the rebase on #95. Hope it is fixed now.

Lorak-mmk commented on scylla-rust-wrapper/src/metadata.rs:
```diff
@@ -0,0 +1,502 @@
+    view_meta: *const CassMaterializedViewMeta,
+) -> *const CassTableMeta {
+    let view_meta = ptr_to_ref(view_meta);
+    view_meta.base_table.as_ptr() as *const CassTableMeta
```
Is the type returned by `as_ptr()` different than `*const CassTableMeta`? If not, remove the cast

Gor027 replied:
The returned type is `*const CassTableMeta`, so the cast is removed.

Lorak-mmk reviewed (changes requested):
I like the overall changes now (apart from few comments I left).
Few issues I have with the commits:
- "Add materialized view metadata ": message is incorrect, there is not RefCell anymore
- "Add impl for functions returning view metadata fields " - this commit should not include moving existing functions between files, that should be a seperate one.

Gor027 commented:
Update:

Rebased on #95 and enabled materialized view metadata tests on GitHub Actions (`SchemaMetadataTest.*View*`).
This PR should be merged only after #95 is merged.

Lorak-mmk commented:
Actually, one more changed is needed - adjustment of the limitations table in README to take into account changes made by this PR

Gor027 commented:
@Lorak-mmk Guess it is good now to be merged.

Gor027 commented:
Also, as you mentioned earlier, a nice optimization maybe not copy the metadata information, but have a pointer to a `ClusterData` object and on each metadata-related function call calculate and return the necessary information. I will open an issue to be discussed in the future.
